File: src/systems/PathfindingSystem.js
export class PathfindingSystem {
  constructor(grid) {
    this.grid = grid;
  }

  findPath(startCol, startRow, endCol, endRow) {
    const sc = Math.round(startCol);
    const sr = Math.round(startRow);
    const ec = Math.round(endCol);
    const er = Math.round(endRow);

    if (!this.grid.inBounds(sc, sr) || !this.grid.inBounds(ec, er)) return null;

    const openSet = [];
    const closedSet = new Set();
    const cameFrom = new Map();
    const gScore = new Map();
    const fScore = new Map();

    const key = (c, r) => `${c},${r}`;
    const startKey = key(sc, sr);
    const endKey = key(ec, er);

    gScore.set(startKey, 0);
    fScore.set(startKey, this._heuristic(sc, sr, ec, er));
    openSet.push({ col: sc, row: sr, f: fScore.get(startKey) });

    let iterations = 0;
    const maxIterations = 2000;

    while (openSet.length > 0 && iterations < maxIterations) {
      iterations++;

      // Get node with lowest fScore
      openSet.sort((a, b) => a.f - b.f);
      const current = openSet.shift();
      const currentKey = key(current.col, current.row);

      if (currentKey === endKey) {
        return this._reconstructPath(cameFrom, current.col, current.row, key);
      }

      closedSet.add(currentKey);

      // 8-directional neighbors
      const dirs = [
        [-1, 0, 1], [1, 0, 1], [0, -1, 1], [0, 1, 1],
        [-1, -1, 1.414], [-1, 1, 1.414], [1, -1, 1.414], [1, 1, 1.414],
      ];

      for (const [dc, dr, cost] of dirs) {
        const nc = current.col + dc;
        const nr = current.row + dr;
        const nKey = key(nc, nr);

        if (closedSet.has(nKey)) continue;
        if (!this.grid.inBounds(nc, nr)) continue;

        const tile = this.grid.getTile(nc, nr);
        // Allow destination tile even if occupied (enemy walks to HQ)
        if (!tile.walkable && nKey !== endKey) continue;

        // Diagonal movement: check that both adjacent cardinal tiles are walkable
        if (dc !== 0 && dr !== 0) {
          const adj1 = this.grid.getTile(current.col + dc, current.row);
          const adj2 = this.grid.getTile(current.col, current.row + dr);
          if ((!adj1 || !adj1.walkable) && (!adj2 || !adj2.walkable)) continue;
        }

        const tentG = gScore.get(currentKey) + cost;

        if (!gScore.has(nKey) || tentG < gScore.get(nKey)) {
          cameFrom.set(nKey, { col: current.col, row: current.row });
          gScore.set(nKey, tentG);
          const f = tentG + this._heuristic(nc, nr, ec, er);
          fScore.set(nKey, f);

          if (!openSet.find(n => key(n.col, n.row) === nKey)) {
            openSet.push({ col: nc, row: nr, f });
          }
        }
      }
    }

    return null; // no path found
  }

  _heuristic(c1, r1, c2, r2) {
    // Chebyshev distance (since we allow diagonal movement)
    const dx = Math.abs(c1 - c2);
    const dy = Math.abs(r1 - r2);
    return Math.max(dx, dy) + (Math.SQRT2 - 1) * Math.min(dx, dy);
  }

  _reconstructPath(cameFrom, endCol, endRow, keyFn) {
    const path = [{ col: endCol, row: endRow }];
    let currentKey = keyFn(endCol, endRow);

    while (cameFrom.has(currentKey)) {
      const prev = cameFrom.get(currentKey);
      path.unshift({ col: prev.col, row: prev.row });
      currentKey = keyFn(prev.col, prev.row);
    }

    return path;
  }
}

File: ios/BiteDefense/Systems/PathfindingSystem.swift
import Foundation

/// A* pathfinding on the occupancy grid. Direct port of `PathfindingSystem.js`.
/// Allows 8-directional movement; the destination tile is always reachable
/// even when occupied (enemies walking onto an HQ tile, for example).
final class PathfindingSystem {
    private unowned let grid: Grid

    init(grid: Grid) {
        self.grid = grid
    }

    struct Step: Equatable { let col: Int; let row: Int }

    func findPath(from start: (col: Int, row: Int),
                  to end:   (col: Int, row: Int),
                  maxIterations: Int = 2000) -> [Step]? {
        let (sc, sr) = start
        let (ec, er) = end
        guard grid.inBounds(col: sc, row: sr),
              grid.inBounds(col: ec, row: er) else { return nil }

        func key(_ c: Int, _ r: Int) -> Int { r * grid.cols + c }
        let endKey = key(ec, er)

        var open: [(col: Int, row: Int, f: Double)] = []
        var closed = Set<Int>()
        var cameFrom: [Int: (col: Int, row: Int)] = [:]
        var g: [Int: Double] = [:]

        g[key(sc, sr)] = 0
        open.append((sc, sr, heuristic(sc, sr, ec, er)))

        var iterations = 0
        while !open.isEmpty, iterations < maxIterations {
            iterations += 1
            // Pop lowest-f node.
            var bestIdx = 0
            for i in 1..<open.count where open[i].f < open[bestIdx].f { bestIdx = i }
            let current = open.remove(at: bestIdx)
            let currentKey = key(current.col, current.row)

            if currentKey == endKey {
                return reconstruct(cameFrom: cameFrom, end: (current.col, current.row))
            }
            closed.insert(currentKey)

            let dirs: [(Int, Int, Double)] = [
                (-1, 0, 1), (1, 0, 1), (0, -1, 1), (0, 1, 1),
                (-1, -1, 1.414), (-1, 1, 1.414), (1, -1, 1.414), (1, 1, 1.414)
            ]
            for (dc, dr, cost) in dirs {
                let nc = current.col + dc
                let nr = current.row + dr
                guard grid.inBounds(col: nc, row: nr) else { continue }
                let nKey = key(nc, nr)
                if closed.contains(nKey) { continue }

                // Destination tile is always enterable (even if occupied).
                let isOccupied = grid.buildingId(at: nc, row: nr) != nil
                if isOccupied && nKey != endKey { continue }

                // For diagonal moves, don't let units squeeze through two corner blockers.
                if dc != 0 && dr != 0 {
                    let a1 = grid.buildingId(at: current.col + dc, row: current.row) != nil
                    let a2 = grid.buildingId(at: current.col, row: current.row + dr) != nil
                    if a1 && a2 { continue }
                }

                let tentG = (g[currentKey] ?? .infinity) + cost
                if tentG < (g[nKey] ?? .infinity) {
                    cameFrom[nKey] = (current.col, current.row)
                    g[nKey] = tentG
                    let f = tentG + heuristic(nc, nr, ec, er)
                    if let i = open.firstIndex(where: { key($0.col, $0.row) == nKey }) {
                        open[i].f = f
                    } else {
                        open.append((nc, nr, f))
                    }
                }
            }
        }
        return nil
    }

    private func heuristic(_ c1: Int, _ r1: Int, _ c2: Int, _ r2: Int) -> Double {
        let dx = Double(abs(c1 - c2))
        let dy = Double(abs(r1 - r2))
        return max(dx, dy) + (sqrt(2.0) - 1) * min(dx, dy)
    }

    private func reconstruct(cameFrom: [Int: (col: Int, row: Int)],
                             end: (col: Int, row: Int)) -> [Step] {
        var path = [Step(col: end.col, row: end.row)]
        func key(_ c: Int, _ r: Int) -> Int { r * grid.cols + c }
        var cur = end
        while let prev = cameFrom[key(cur.col, cur.row)] {
            path.insert(Step(col: prev.col, row: prev.row), at: 0)
            cur = prev
        }
        return path
    }
}
